EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c

```diff
@@ -0,0 +1,77 @@
+#include "UpdateChargeAllowanceTask.h"
+
+static const uint32_t CHARGE_ALLOWANCE_UPDATE_FREQ = 50; // Every 100ms
+// These values are subject to change based on what regulations say
+// Ask Dan if any issues
+static const float MAX_CELL_VOLTAGE = 3.8;
+static const float MIN_CELL_VOLTAGE = 3.2;
+typedef enum ChargeAllowanceState {VOLTAGE_IN_RANGE,
+                                   OUT_OF_RANGE
+                                  } ChargeAllowanceState;
+
+void updateChargeAllowanceTask(void const* arg)
+{
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+    ChargeAllowanceState maxCellVoltageChargeAllowanceState = VOLTAGE_IN_RANGE;
```

Comment:
The logic for this file can be simpler, try something like this.

```C
void updateChargeAllowanceTask(void const* arg)
{
    // One time osDelayUntil initialization
    uint32_t prevWakeTime = osKernelSysTick();

    for (;;)
    {
        osDelayUntil(&prevWakeTime, CHARGE_ALLOWANCE_UPDATE_FREQ);

        /* Orion BMS sends voltages relating to two DIFFERENT cells: the cell with the highest voltage
           and the cell with the lowest voltage. Must check if the voltage of the cell with the highest
           voltage is too high or in range and if the cell with the lowest voltage is too low or in range.
        */
        uint8_t voltageInRange = 1;
        uint8_t allowCharge = 1;
        uint8_t allowDischarge = 1;

        if ((double)orionStatus.maxCellVoltage > MAX_CELL_VOLTAGE * 0.8) // Will have the cutoff to be 20% below
        {
            // Turn off charge contactor
            HAL_GPIO_WritePin(CONTACTOR_ENABLE2_GPIO_Port, CONTACTOR_ENABLE2_Pin, GPIO_PIN_RESET);
            allowCharge = 0;
            voltageInRange = 0;
        }

        if ((double)orionStatus.minCellVoltage < MIN_CELL_VOLTAGE * 1.2) // Will have the cutoff to be 20% above
        {
            // Turn off High voltage enable and discharge contactor
            HAL_GPIO_WritePin(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin, GPIO_PIN_RESET);
            HAL_GPIO_WritePin(CONTACTOR_ENABLE3_GPIO_Port, CONTACTOR_ENABLE3_Pin, GPIO_PIN_RESET);
            allowDischarge = 0;
            voltageInRange = 0;
        }

        // Setting the allowance of charge and charge/discharge contactor state for auxStatus
        if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
        {
            continue;
        }

        auxStatus.allowCharge = allowCharge;
        auxStatus.allowDischarge = allowDischarge;
        orionStatus.batteryVoltagesInRange = voltageInRange;

        osMutexRelease(auxStatus.auxStatusMutex);
    }
}
```

Overall I would say this file should just update some internal state. And let `setContactorTask()` actually physically deal with the contactors.

Also a new question in my head, you used the 20% safety factors, did electrical say this was ok? Because of how voltage works, they're usually in a curve, so 20% voltage isn't actually 20% battery usage, it will probably be more like 40% battery usage. Some example voltage curves (SOC curves), https://chargedevs.com/wp-content/uploads/2013/09/SOC-Measurments-figure11.png.

Also what task is supposed to be controlling the `HV_ENABLE_GPIO`? You turn it off here, but is it possible that it will get turned on again somewhere else?

Reply:
The 20% was directly from Dan and Logan. The reason I'm also messing with the contactors here is because they said there needs to be another, much faster task, that can shut everything down ASAP. SetContactors is a much slower task. HV_ENABLE is turned on at the very end of setContactors. Could possibly make another task just for that, but it's one line so I didn't think it was necessary